NASA-IMPACT/admg-backend
# Website and Alias Forms (and feature friends)

# What was built
Too many things, in short.  This became a mild Frankenstein PR and for that I apologize. 

* adding a "related fields" section of forms that includes Aliases for 'Campaign', 'Platform', 'Deployment', 'Instrument', and 'PartnerOrg' forms and CampaignWebsites for Campaign pages.
* added the library Select2 to select fields for nicer multi-select dropdowns
* small requests from ADMG (flip order of IOP and SE on campaign details page, combining the CMR and DOI fields on the DOI tables, and probably some other things I can't remember)

# How it was built
### Aliases and Websites Forms
The related fields section is an HTML snippet that gets referenced from the campaign details page.  Data is input into the snippet from `views.py` in the `ChangeUpdateView` class using the method `get_related_fields` which specifies the appropriate related fields.  

Websites display the CampaignWebsite, not the website class.  CampaignWebsite should pre-populate with the campaign that you just navigated from.

There was some craziness with creating CampaignWebsite and Alias forms since those two entities didn't have short names.  This was addressed using the `formfield_callback` function in `mixins.py` to specify custom params by field instance type.

### Not built
* Adding the popup into the campaignwebsite form to allow the user to create a new website.
* The back buttons are still absolutely terrible.  I'm hoping to fix this one in the morning.
* Websites are listed by title in the Campaign form and I have a feeling the titles will actually be a lot longer than the random text I was testing with.  The point is that the websites overflow the side column which looks super unpolished.

# How to test it
Go try out the user flow to create new websites and aliases.  Aliases should be available on all the major types (Campaign, Platform, Instrument, PartnerOrg).
